File: src/main/java/net/yakclient/mixin/internal/instruction/adapter/FieldSelfInsnAdapter.java
package net.yakclient.mixin.internal.instruction.adapter;

import net.yakclient.mixin.internal.instruction.InsnAdapter;
import net.yakclient.mixin.internal.instruction.Instruction;
import org.objectweb.asm.Opcodes;
import org.objectweb.asm.tree.AbstractInsnNode;
import org.objectweb.asm.tree.FieldInsnNode;

public class FieldSelfInsnAdapter extends InsnAdapter {
    private final String clsTo;
    private final String clsFrom;

    public FieldSelfInsnAdapter(InsnAdapter adapter, String clsTo, String clsFrom) {
        super(adapter);
        this.clsTo = clsTo;
        this.clsFrom = clsFrom;
    }

    public FieldSelfInsnAdapter(String clsTo, String clsFrom) {
        this.clsTo = clsTo;
        this.clsFrom = clsFrom;
    }

    @Override
    public Instruction adapt(Instruction instruction) {
        for (AbstractInsnNode insn : instruction.getInstructions()) {
            if (insn instanceof FieldInsnNode &&
                    insn.getOpcode() != Opcodes.GETSTATIC &&
                    ((FieldInsnNode) insn).owner.equals(this.clsFrom))
                ((FieldInsnNode) insn).owner = this.clsTo;
        }
        return super.adapt(instruction);
    }
}

File: mixin-base/src/main/kotlin/net/yakclient/mixin/base/internal/instruction/adapter/FieldSelfInsnAdapter.kt
package net.yakclient.mixin.base.internal.instruction.adapter

import net.yakclient.mixin.base.internal.instruction.Instruction
import org.objectweb.asm.Opcodes
import org.objectweb.asm.tree.FieldInsnNode

class FieldSelfInsnAdapter : InsnAdapter {
    private val clsTo: String
    private val clsFrom: String

    constructor(adapter: InsnAdapter?, clsTo: String, clsFrom: String) : super(adapter) {
        this.clsTo = clsTo.replace(".", "/")
        this.clsFrom = clsFrom.replace(".", "/")
    }

    override fun adapt(instruction: Instruction): Instruction {
        for (insn in instruction.insn) {
            if (insn is FieldInsnNode && insn.getOpcode() != Opcodes.GETSTATIC && insn.owner == clsFrom) insn.owner =
                clsTo
        }
        return super.adapt(instruction)
    }
}